26.2 Support

### Confirmation

- [x] My issue isn't already found on the Issue tracker.
- [x] My issue is about **PlaceholderAPI** and not any expansion or external plugin.
- [x] The issue isn't already fixed in a Spigot Release or Development Build.
- [x] The [Common Issues](https://wiki.placeholderapi.com/common-issues/) page doesn't mention this issue.

### Type

Plugin Bug

### What happens?

[18:08:01 ERROR]: [ModernPluginLoadingStrategy] Could not load plugin 'PlaceholderAPI-2.12.2.jar' in folder 'plugins'
org.bukkit.plugin.InvalidPluginException: java.lang.ExceptionInInitializerError
        at io.papermc.paper.plugin.provider.type.spigot.SpigotPluginProvider.createInstance(SpigotPluginProvider.java:129) ~[paper-26.2.jar:26.2-10-5979742]
        at io.papermc.paper.plugin.provider.type.spigot.SpigotPluginProvider.createInstance(SpigotPluginProvider.java:35) ~[paper-26.2.jar:26.2-10-5979742]
        at io.papermc.paper.plugin.entrypoint.strategy.modern.ModernPluginLoadingStrategy.loadProviders(ModernPluginLoadingStrategy.java:116) ~[paper-26.2.jar:26.2-10-5979742]
        at io.papermc.paper.plugin.storage.SimpleProviderStorage.enter(SimpleProviderStorage.java:38) ~[paper-26.2.jar:26.2-10-5979742]
        at io.papermc.paper.plugin.entrypoint.LaunchEntryPointHandler.enter(LaunchEntryPointHandler.java:39) ~[paper-26.2.jar:26.2-10-5979742]
        at org.bukkit.craftbukkit.CraftServer.loadPlugins(CraftServer.java:549) ~[paper-26.2.jar:26.2-10-5979742]
        at net.minecraft.server.dedicated.DedicatedServer.initServer(DedicatedServer.java:297) ~[paper-26.2.jar:26.2-10-5979742]
        at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1290) ~[paper-26.2.jar:26.2-10-5979742]
        at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:302) ~[paper-26.2.jar:26.2-10-5979742]
        at java.base/java.lang.Thread.run(Thread.java:1474) ~[?:?]
Caused by: java.lang.ExceptionInInitializerError
        at java.base/java.lang.Class.forName0(Native Method) ~[?:?]
        at java.base/java.lang.Class.forName(Class.java:547) ~[?:?]
        at org.bukkit.plugin.java.PluginClassLoader.<init>(PluginClassLoader.java:82) ~[paper-api-26.2.build.10-alpha.jar:?]
        at io.papermc.paper.plugin.provider.type.spigot.SpigotPluginProvider.createInstance(SpigotPluginProvider.java:125) ~[paper-26.2.jar:26.2-10-5979742]
        ... 9 more
Caused by: java.lang.NumberFormatException: For input string: "b"
        at java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:67) ~[?:?]
        at java.base/java.lang.Integer.parseInt(Integer.java:565) ~[?:?]
        at java.base/java.lang.Integer.parseInt(Integer.java:662) ~[?:?]
        at PlaceholderAPI-2.12.2.jar//me.clip.placeholderapi.PlaceholderAPIPlugin.<clinit>(PlaceholderAPIPlugin.java:71) ~[?:?]
        ... 13 more

### Expected Behaviour

no error

### How to Reproduce

Starting Paper 26.2 Server with PlaceholderAPI 2.12.2

### `/papi dump` Output

/

### Console Log

_No response_

### Errors

_No response_

### Additional Info

_No response_

--- a/src/main/java/me/clip/placeholderapi/PlaceholderAPIPlugin.java
+++ b/src/main/java/me/clip/placeholderapi/PlaceholderAPIPlugin.java
@@ -60,7 +60,8 @@ public final class PlaceholderAPIPlugin extends JavaPlugin {
     private static PlaceholderAPIPlugin instance;
 
     static {
-        String version = Bukkit.getServer().getBukkitVersion().split("-")[0];
+        String version = normalizeBukkitVersion(Bukkit.getServer().getBukkitVersion());
+
         String suffix;
         if (version.chars()
                 .filter(c -> c == '.')
@@ -68,7 +69,17 @@ public final class PlaceholderAPIPlugin extends JavaPlugin {
             suffix = "R1";
             version = 'v' + version.replace('.', '_') + '_' + suffix;
         } else {
-            int minor = Integer.parseInt(version.split("\\.")[2].charAt(0) + "");
+            final String[] versionParts = version.split("\\.");
+
+            int minor = 1;
+            if (versionParts.length > 2 && !versionParts[2].isEmpty()) {
+                try {
+                    minor = Integer.parseInt(versionParts[2].charAt(0) + "");
+                } catch (final NumberFormatException ignored) {
+                    minor = 1;
+                }
+            }
+
             version = 'v' + version.replace('.', '_').replace("_" + minor, "") + '_' + "R" + (minor - 1);
         }
 
@@ -294,4 +305,15 @@ public final class PlaceholderAPIPlugin extends JavaPlugin {
         }
     }
 
+    @NotNull
+    private static String normalizeBukkitVersion(@NotNull final String bukkitVersion) {
+        String version = bukkitVersion.split("-", 2)[0];
+
+        final int paperBuildMetadataIndex = version.indexOf(".build.");
+        if (paperBuildMetadataIndex != -1) {
+            version = version.substring(0, paperBuildMetadataIndex);
+        }
+
+        return version;
+    }
 }

--- a/src/main/java/me/clip/placeholderapi/PlaceholderAPIPlugin.java
+++ b/src/main/java/me/clip/placeholderapi/PlaceholderAPIPlugin.java
@@ -60,28 +60,8 @@ public final class PlaceholderAPIPlugin extends JavaPlugin {
     private static PlaceholderAPIPlugin instance;
 
     static {
-        String version = normalizeBukkitVersion(Bukkit.getServer().getBukkitVersion());
-
-        String suffix;
-        if (version.chars()
-                .filter(c -> c == '.')
-                .count() == 1) {
-            suffix = "R1";
-            version = 'v' + version.replace('.', '_') + '_' + suffix;
-        } else {
-            final String[] versionParts = version.split("\\.");
-
-            int minor = 1;
-            if (versionParts.length > 2 && !versionParts[2].isEmpty()) {
-                try {
-                    minor = Integer.parseInt(versionParts[2].charAt(0) + "");
-                } catch (final NumberFormatException ignored) {
-                    minor = 1;
-                }
-            }
-
-            version = 'v' + version.replace('.', '_').replace("_" + minor, "") + '_' + "R" + (minor - 1);
-        }
+        final String version = ServerVersionResolver.resolve(
+                Bukkit.getServer().getBukkitVersion()).getLegacyVersion();
 
         boolean isSpigot;
         try {
@@ -305,15 +285,4 @@ public final class PlaceholderAPIPlugin extends JavaPlugin {
         }
     }
 
-    @NotNull
-    private static String normalizeBukkitVersion(@NotNull final String bukkitVersion) {
-        String version = bukkitVersion.split("-", 2)[0];
-
-        final int paperBuildMetadataIndex = version.indexOf(".build.");
-        if (paperBuildMetadataIndex != -1) {
-            version = version.substring(0, paperBuildMetadataIndex);
-        }
-
-        return version;
-    }
 }

--- a/src/main/java/me/clip/placeholderapi/ServerVersionResolver.java
+++ b/src/main/java/me/clip/placeholderapi/ServerVersionResolver.java
@@ -1,0 +1,133 @@
+/*
+ * This file is part of PlaceholderAPI
+ *
+ * PlaceholderAPI
+ * Copyright (c) 2015 - 2026 PlaceholderAPI Team
+ *
+ * PlaceholderAPI free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PlaceholderAPI is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package me.clip.placeholderapi;
+
+import java.util.OptionalInt;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+final class ServerVersionResolver {
+
+    private static final String PAPER_BUILD_INFO_PROVIDER =
+            "me.clip.placeholderapi.PaperServerBuildInfoProvider";
+
+    private ServerVersionResolver() {
+    }
+
+    @NotNull
+    static ServerBuild resolve(@NotNull final String bukkitVersion) {
+        return resolve(bukkitVersion, resolvePaperBuildInfo());
+    }
+
+    @NotNull
+    static ServerBuild resolve(@NotNull final String bukkitVersion,
+                               @Nullable final ServerBuild paperBuildInfo) {
+        if (paperBuildInfo != null) {
+            return paperBuildInfo;
+        }
+
+        return new ServerBuild(normalizeBukkitVersion(bukkitVersion), OptionalInt.empty());
+    }
+
+    @Nullable
+    private static ServerBuild resolvePaperBuildInfo() {
+        try {
+            final Class<?> provider = Class.forName(
+                    PAPER_BUILD_INFO_PROVIDER,
+                    true,
+                    ServerVersionResolver.class.getClassLoader()
+            );
+            final Object result = provider.getMethod("get").invoke(null);
+
+            if (result instanceof ServerBuild) {
+                return (ServerBuild) result;
+            }
+        } catch (final ReflectiveOperationException | LinkageError ignored) {
+            // Paper is unavailable or exposes an incompatible ServerBuildInfo API.
+        }
+
+        return null;
+    }
+
+    @NotNull
+    static String normalizeBukkitVersion(@NotNull final String bukkitVersion) {
+        String version = bukkitVersion.split("-", 2)[0];
+
+        // Some Paper development versions include build metadata in the Bukkit version.
+        // Keep this fallback for Paper forks that do not expose ServerBuildInfo.
+        final int paperBuildMetadataIndex = version.indexOf(".build.");
+        if (paperBuildMetadataIndex != -1) {
+            version = version.substring(0, paperBuildMetadataIndex);
+        }
+
+        return version;
+    }
+
+    static final class ServerBuild {
+
+        @NotNull
+        private final String minecraftVersionId;
+        @NotNull
+        private final OptionalInt buildNumber;
+
+        ServerBuild(@NotNull final String minecraftVersionId,
+                    @NotNull final OptionalInt buildNumber) {
+            this.minecraftVersionId = minecraftVersionId;
+            this.buildNumber = buildNumber;
+        }
+
+        @NotNull
+        String getMinecraftVersionId() {
+            return minecraftVersionId;
+        }
+
+        @NotNull
+        OptionalInt getBuildNumber() {
+            return buildNumber;
+        }
+
+        @NotNull
+        String getLegacyVersion() {
+            String version = minecraftVersionId;
+
+            if (version.chars()
+                    .filter(c -> c == '.')
+                    .count() == 1) {
+                return 'v' + version.replace('.', '_') + "_R1";
+            }
+
+            final String[] versionParts = version.split("\\.");
+            int minor = 1;
+
+            if (versionParts.length > 2 && !versionParts[2].isEmpty()) {
+                try {
+                    minor = Integer.parseInt(versionParts[2].charAt(0) + "");
+                } catch (final NumberFormatException ignored) {
+                    minor = 1;
+                }
+            }
+
+            return 'v' + version.replace('.', '_').replace("_" + minor, "")
+                    + "_R" + (minor - 1);
+        }
+    }
+}

--- a/src/paper/java/me/clip/placeholderapi/PaperServerBuildInfoProvider.java
+++ b/src/paper/java/me/clip/placeholderapi/PaperServerBuildInfoProvider.java
@@ -1,0 +1,42 @@
+/*
+ * This file is part of PlaceholderAPI
+ *
+ * PlaceholderAPI
+ * Copyright (c) 2015 - 2026 PlaceholderAPI Team
+ *
+ * PlaceholderAPI free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PlaceholderAPI is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package me.clip.placeholderapi;
+
+import io.papermc.paper.ServerBuildInfo;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+
+@ApiStatus.Internal
+public final class PaperServerBuildInfoProvider {
+
+    private PaperServerBuildInfoProvider() {
+    }
+
+    @NotNull
+    public static ServerVersionResolver.ServerBuild get() {
+        final ServerBuildInfo buildInfo = ServerBuildInfo.buildInfo();
+
+        return new ServerVersionResolver.ServerBuild(
+                buildInfo.minecraftVersionId(),
+                buildInfo.buildNumber()
+        );
+    }
+}

--- a/src/test/java/me/clip/placeholderapi/ServerVersionResolverTest.java
+++ b/src/test/java/me/clip/placeholderapi/ServerVersionResolverTest.java
@@ -1,0 +1,64 @@
+/*
+ * This file is part of PlaceholderAPI
+ *
+ * PlaceholderAPI
+ * Copyright (c) 2015 - 2026 PlaceholderAPI Team
+ *
+ * PlaceholderAPI free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PlaceholderAPI is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package me.clip.placeholderapi;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertSame;
+
+import java.util.OptionalInt;
+
+import org.junit.jupiter.api.Test;
+
+public final class ServerVersionResolverTest {
+
+    @Test
+    void usesPaperBuildInfoWhenAvailable() {
+        final ServerVersionResolver.ServerBuild paperBuild =
+                new ServerVersionResolver.ServerBuild("26.2", OptionalInt.of(10));
+
+        final ServerVersionResolver.ServerBuild result =
+                ServerVersionResolver.resolve("26.2.build.10-alpha", paperBuild);
+
+        assertSame(paperBuild, result);
+        assertEquals("26.2", result.getMinecraftVersionId());
+        assertEquals(OptionalInt.of(10), result.getBuildNumber());
+        assertEquals("v26_2_R1", result.getLegacyVersion());
+    }
+
+    @Test
+    void stripsPaperBuildMetadataWhenServerBuildInfoIsUnavailable() {
+        final ServerVersionResolver.ServerBuild result =
+                ServerVersionResolver.resolve("26.2.build.10-alpha", null);
+
+        assertEquals("26.2", result.getMinecraftVersionId());
+        assertEquals(OptionalInt.empty(), result.getBuildNumber());
+        assertEquals("v26_2_R1", result.getLegacyVersion());
+    }
+
+    @Test
+    void preservesTraditionalBukkitVersions() {
+        final ServerVersionResolver.ServerBuild result =
+                ServerVersionResolver.resolve("1.21.11-R0.1-SNAPSHOT", null);
+
+        assertEquals("1.21.11", result.getMinecraftVersionId());
+        assertEquals(OptionalInt.empty(), result.getBuildNumber());
+    }
+}

